tests/tape/test_tape.py

```diff
@@ -1466,3 +1466,225 @@ def test_gate_tape():
         assert qml.tape.get_active_tape() is tape1
 
     assert qml.tape.get_active_tape() is None
+
+
+class TestHashing:
+    """Test for tape hashing"""
+
+    @pytest.mark.parametrize(
+        "m",
+        [
+            qml.expval(qml.PauliZ(0)),
+            qml.state(),
+            qml.probs(wires=0),
+            qml.density_matrix(wires=0),
+            qml.var(qml.PauliY(0)),
+        ],
+    )
+    def test_identical(self, m):
+        """Tests that the circuit hash of identical circuits are identical"""
+        a = 0.3
+        b = 0.2
+
+        with qml.tape.QuantumTape() as tape1:
+            qml.RX(a, wires=[0])
+            qml.RY(b, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.apply(m)
+
+        with qml.tape.QuantumTape() as tape2:
+            qml.RX(a, wires=[0])
+            qml.RY(b, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.apply(m)
+
+        assert tape1.hash == tape2.hash
+
+    def test_identical_numeric(self):
+        """Tests that the circuit hash of identical circuits are identical
+        even though the datatype of the arguments may differ"""
+        a = 0.3
+        b = 0.2
+
+        with qml.tape.QuantumTape() as tape1:
+            qml.RX(a, wires=[0])
+            qml.RY(b, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        with qml.tape.QuantumTape() as tape2:
+            qml.RX(np.array(a), wires=[0])
+            qml.RY(np.array(b), wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        assert tape1.hash == tape2.hash
+
+    def test_different_wires(self):
+        """Tests that the circuit hash of circuits with the same operations
+        on different wires have different hashes"""
+        a = 0.3
+        b = 0.2
+
+        with qml.tape.QuantumTape() as tape1:
+            qml.RX(a, wires=[1])
+            qml.RY(b, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        with qml.tape.QuantumTape() as tape2:
+            qml.RX(np.array(a), wires=[0])
+            qml.RY(np.array(b), wires=[1])
```

Comment:
Is the `np.array()` left over from the previous test? Though I guess it doesn't matter because the hash should be the same.

Reply:
oh this was semi-intentional - I was trying to ensure that the datatype of the parameter doesn't affect hashing 😆